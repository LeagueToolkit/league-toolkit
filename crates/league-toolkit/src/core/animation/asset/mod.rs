mod compressed;

pub use compressed::*;

mod uncompressed;

pub use uncompressed::*;

pub mod error;
mod error_metric;

pub use error::*;

use error::AssetParseError::UnknownAssetType;
use std::io;
use std::io::{Read, Seek, SeekFrom};

#[derive(Clone, Debug)]
pub enum AnimationAsset {
    Uncompressed(Uncompressed),
    Compressed(Compressed),
}

pub enum AnimationAssetType {
    Uncompressed,
    Compressed,
    Unknown,
}

impl AnimationAsset {
    pub fn from_reader<R: Read + Seek + ?Sized>(reader: &mut R) -> error::Result<AnimationAsset> {
        use crate::util::ReaderExt as _;
        use byteorder::{ReadBytesExt as _, LE};
        let asset_type = Self::identify_from_reader(reader)?;
        reader.seek(SeekFrom::Start(0))?;
        match asset_type {
            AnimationAssetType::Uncompressed => {
                Uncompressed::from_reader(reader).map(Self::Uncompressed)
            }
            AnimationAssetType::Compressed => Compressed::from_reader(reader).map(Self::Compressed),
            AnimationAssetType::Unknown => Err(UnknownAssetType),
        }
    }

    pub fn identify_from_reader<R: Read + ?Sized>(
        reader: &mut R,
    ) -> io::Result<AnimationAssetType> {
        let mut magic = [0_u8; 8];
        reader.read_exact(&mut magic)?;

        Ok(match &magic {
            b"r3d2anmd" => AnimationAssetType::Uncompressed,
            b"r3d2canm" => AnimationAssetType::Compressed,
            _ => AnimationAssetType::Unknown,
        })
    }
}
