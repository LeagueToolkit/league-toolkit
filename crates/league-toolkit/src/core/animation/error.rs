#[derive(Debug, thiserror::Error)]
pub enum ParseError {
    #[error("Invalid file signature")]
    InvalidFileSignature,
    #[error("Invalid file version '{0}'")]
    InvalidFileVersion(u32),
    #[error("Invalid '{0}' - got '{1}'")]
    InvalidField(&'static str, String),
    #[error("IO Error - {0}")]
    ReaderError(#[from] std::io::Error),
    #[error("UTF-8 Error - {0}")]
    Utf8Error(#[from] std::str::Utf8Error),
}

pub type Result<T> = core::result::Result<T, ParseError>;
