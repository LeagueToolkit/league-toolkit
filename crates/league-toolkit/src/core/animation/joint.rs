use std::io;
use std::io::{Read, Seek, SeekFrom};
use byteorder::ReadBytesExt;
use glam::{Mat4, Quat, Vec3};
use crate::util::ReaderExt;

#[derive(Debug, Clone, PartialEq)]
pub struct Joint {
    name: String,
    flags: u16,
    id: i16,
    parent_id: i16,
    radius: f32,
    local_transform: Mat4,
    local_translation: Vec3,
    local_scale: Vec3,
    local_rotation: Quat,
    inverse_bind_transform: Mat4,
    inverse_bind_translation: Vec3,
    inverse_bind_scale: Vec3,
    inverse_bind_rotation: Quat,

}

impl Joint {
    pub fn new(name: String, flags: u16, id: i16, parent_id: i16, radius: f32, local_transform: Mat4, inverse_bind_transform: Mat4) -> Self {
        let (local_scale, local_rotation, local_translation) = local_transform.to_scale_rotation_translation();
        let (inverse_bind_scale, inverse_bind_rotation, inverse_bind_translation) = inverse_bind_transform.to_scale_rotation_translation();

        Self {
            name,
            flags,
            id,
            parent_id,
            radius,
            local_transform,
            local_translation,
            local_scale,
            local_rotation,
            inverse_bind_transform,
            inverse_bind_translation,
            inverse_bind_scale,
            inverse_bind_rotation,
        }
    }

    pub fn from_reader<R: Read + Seek + ?Sized>(reader: &mut R) -> io::Result<Self> {
        use byteorder::{ReadBytesExt as _, LE};
        use crate::util::ReaderExt as _;

        let flags = reader.read_u16::<LE>()?;
        let id = reader.read_i16::<LE>()?;
        let parent_id = reader.read_i16::<LE>()?;
        reader.read_i16::<LE>()?; // padding
        let name_hash = reader.read_u32::<LE>()?;
        let radius = reader.read_f32::<LE>()?;

        let local_translation = reader.read_vec3::<LE>()?;
        let local_scale = reader.read_vec3::<LE>()?;
        let local_rotation = reader.read_quat::<LE>()?.normalize();

        let inverse_bind_translation = reader.read_vec3::<LE>()?;
        let inverse_bind_scale = reader.read_vec3::<LE>()?;
        let inverse_bind_rotation = reader.read_quat::<LE>()?;

        let name_off = reader.read_i32::<LE>()?;
        let return_pos = reader.stream_position()?;

        reader.seek(SeekFrom::Current(-4 + name_off as i64))?;
        let name = reader.read_str_until_nul()?;
        reader.seek(SeekFrom::Start(return_pos))?;

        Ok(Self {
            name,
            flags,
            id,
            parent_id,
            radius,
            local_transform: Mat4::from_scale_rotation_translation(local_scale, local_rotation.normalize(), local_translation),
            local_translation,
            local_scale,
            local_rotation,
            inverse_bind_transform: Mat4::from_scale_rotation_translation(inverse_bind_scale, inverse_bind_rotation.normalize(), inverse_bind_translation),
            inverse_bind_translation,
            inverse_bind_scale,
            inverse_bind_rotation,
        })
    }
    pub fn name(&self) -> &str {
        &self.name
    }
    pub fn flags(&self) -> u16 {
        self.flags
    }
    pub fn id(&self) -> i16 {
        self.id
    }
    pub fn parent_id(&self) -> i16 {
        self.parent_id
    }
    pub fn radius(&self) -> f32 {
        self.radius
    }
    pub fn local_transform(&self) -> Mat4 {
        self.local_transform
    }
    pub fn local_translation(&self) -> Vec3 {
        self.local_translation
    }
    pub fn local_scale(&self) -> Vec3 {
        self.local_scale
    }
    pub fn local_rotation(&self) -> Quat {
        self.local_rotation
    }
    pub fn inverse_bind_transform(&self) -> Mat4 {
        self.inverse_bind_transform
    }
    pub fn inverse_bind_translation(&self) -> Vec3 {
        self.inverse_bind_translation
    }
    pub fn inverse_bind_scale(&self) -> Vec3 {
        self.inverse_bind_scale
    }
    pub fn inverse_bind_rotation(&self) -> Quat {
        self.inverse_bind_rotation
    }
}


#[derive(Debug, Clone, PartialEq)]
pub struct LegacyJoint {
    name: String,
    id: i16,
    parent_id: i16,
    radius: f32,
    global_transform: Mat4,
}

impl LegacyJoint {
    pub fn from_reader<R: Read + ?Sized>(reader: &mut R, id: i16) -> io::Result<Self> {
        use byteorder::{ReadBytesExt as _, LE};
        use crate::util::ReaderExt as _;

        let name = reader.read_padded_string::<LE, 32>().expect("FIXME: better error here");
        let parent_id = reader.read_i32::<LE>()? as i16;
        let radius = reader.read_f32::<LE>()?;
        let mut transform = [[0.0; 4]; 4];
        transform[3][3] = 1.0;
        for i in 0..3 {
            for j in 0..4 {
                transform[j][i] = reader.read_f32::<LE>()?;
            }
        }

        Ok(Self {
            name,
            id,
            parent_id,
            radius,
            global_transform: Mat4::from_cols_array_2d(&transform),
        })
    }
}