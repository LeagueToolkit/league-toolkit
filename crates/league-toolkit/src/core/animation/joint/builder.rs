use glam::Mat4;

pub struct Builder {
    name: String,
    flags: u16,
    is_influence: bool,
    parent: Option<()>,
    radius: f32,
    local_transform: Mat4,
    inverse_bind_transform: Mat4,
    children: Vec<()>,
}

impl Builder {
    pub fn new(name: String) -> Self {
        Self {
            name,
            flags: 0,
            is_influence: false,
            parent: None,
            radius: 2.1,
            local_transform: Default::default(),
            inverse_bind_transform: Default::default(),
            children: vec![],
        }
    }

    pub fn with_flags(mut self, flags: u16) -> Self {
        self.flags = flags;
        self
    }

    pub fn with_influence(mut self, is_influence: bool) -> Self {
        self.is_influence = is_influence;
        self
    }
    pub fn with_local_transform(mut self, is_influence: bool) -> Self {
        self.is_influence = is_influence;
        self
    }
    pub fn with_inverse_bind_transform(mut self, is_influence: bool) -> Self {
        self.is_influence = is_influence;
        self
    }

    pub fn new_child(&mut self, name: String) -> Self {
        let child = Self::new(name);
        // self.children.push(child);
        child
    }
}