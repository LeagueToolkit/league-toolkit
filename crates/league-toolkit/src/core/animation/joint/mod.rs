use std::io::{Read, Seek};
use byteorder::ReadBytesExt;
use glam::{Mat4, Quat, Vec3};
use crate::util::ReaderExt;

pub mod legacy;
mod read;
mod write;
mod builder;

pub use builder::Builder;

#[derive(Debug, Clone, PartialEq)]
pub struct Joint {
    name: String,
    flags: u16,
    id: i16,
    parent_id: i16,
    radius: f32,
    local_transform: Mat4,
    local_translation: Vec3,
    local_scale: Vec3,
    local_rotation: Quat,
    inverse_bind_transform: Mat4,
    inverse_bind_translation: Vec3,
    inverse_bind_scale: Vec3,
    inverse_bind_rotation: Quat,

}

impl Joint {
    pub fn builder(name: impl Into<String>) -> Builder {
        Builder::new(name)
    }
    pub fn name(&self) -> &str {
        &self.name
    }
    pub fn flags(&self) -> u16 {
        self.flags
    }
    pub fn id(&self) -> i16 {
        self.id
    }
    pub fn parent_id(&self) -> i16 {
        self.parent_id
    }
    pub fn radius(&self) -> f32 {
        self.radius
    }
    pub fn local_transform(&self) -> Mat4 {
        self.local_transform
    }
    pub fn local_translation(&self) -> Vec3 {
        self.local_translation
    }
    pub fn local_scale(&self) -> Vec3 {
        self.local_scale
    }
    pub fn local_rotation(&self) -> Quat {
        self.local_rotation
    }
    pub fn inverse_bind_transform(&self) -> Mat4 {
        self.inverse_bind_transform
    }
    pub fn inverse_bind_translation(&self) -> Vec3 {
        self.inverse_bind_translation
    }
    pub fn inverse_bind_scale(&self) -> Vec3 {
        self.inverse_bind_scale
    }
    pub fn inverse_bind_rotation(&self) -> Quat {
        self.inverse_bind_rotation
    }
}
