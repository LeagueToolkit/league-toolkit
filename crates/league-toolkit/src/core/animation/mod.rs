pub mod joint;

pub use joint::*;

pub mod error;

pub use error::*;

pub mod asset;
pub mod rig;

pub use asset::{AnimationAsset, AnimationAssetType, AssetParseError, Compressed, Uncompressed};

pub use rig::*;
