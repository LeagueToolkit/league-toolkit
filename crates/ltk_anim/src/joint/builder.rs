use crate::Joint;
use glam::Mat4;

#[derive(Clone, Debug)]
pub struct Builder {
    pub name: String,
    pub flags: u16,
    pub is_influence: bool,
    pub radius: f32,
    pub local_transform: Mat4,
    pub inverse_bind_transform: Mat4,
    pub children: Vec<Box<Builder>>,
}

impl Builder {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            flags: 0,
            is_influence: false,
            radius: 2.1,
            local_transform: Default::default(),
            inverse_bind_transform: Default::default(),
            children: vec![],
        }
    }

    pub fn with_flags(mut self, flags: u16) -> Self {
        self.flags = flags;
        self
    }

    pub fn with_influence(mut self, is_influence: bool) -> Self {
        self.is_influence = is_influence;
        self
    }
    pub fn with_local_transform(mut self, local_transform: Mat4) -> Self {
        self.local_transform = local_transform;
        self
    }
    pub fn with_inverse_bind_transform(mut self, inverse_bind_transform: Mat4) -> Self {
        self.inverse_bind_transform = inverse_bind_transform;
        self
    }

    pub fn with_children<I: IntoIterator<Item = impl Into<Box<Builder>>>>(
        mut self,
        children: I,
    ) -> Self {
        self.add_children(children);
        self
    }
    pub fn add_children<I: IntoIterator<Item = impl Into<Box<Builder>>>>(&mut self, children: I) {
        self.children.extend(children.into_iter().map(|c| c.into()));
    }

    pub fn add_child(&mut self, child: Box<Builder>) {
        self.children.push(child);
    }

    pub fn build(self, id: i16, parent_id: i16) -> (Joint, Vec<Box<Builder>>) {
        (
            Joint::new(
                self.name,
                self.flags,
                id,
                parent_id,
                self.radius,
                self.local_transform,
                self.inverse_bind_transform,
            ),
            self.children,
        )
    }
}
