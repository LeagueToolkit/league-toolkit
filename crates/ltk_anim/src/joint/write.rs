use crate::Joint;
use byteorder::{WriteBytesExt, LE};
use elf_hash::hash;
use io_ext::WriterExt;
use std::io;
use std::io::{Seek, Write};

impl Joint {
    pub fn to_writer<W: Write + Seek + ?Sized>(
        &self,
        writer: &mut W,
        name_off: u64,
    ) -> io::Result<()> {
        writer.write_u16::<LE>(self.flags)?;
        writer.write_i16::<LE>(self.id)?;
        writer.write_i16::<LE>(self.parent_id)?;

        writer.write_i16::<LE>(0)?; // padding

        writer.write_u32::<LE>(hash::elf(&self.name) as u32)?;
        writer.write_f32::<LE>(self.radius)?;

        writer.write_vec3::<LE>(&self.local_translation)?;
        writer.write_vec3::<LE>(&self.local_scale)?;
        writer.write_quat::<LE>(&self.local_rotation)?;

        writer.write_vec3::<LE>(&self.inverse_bind_translation)?;
        writer.write_vec3::<LE>(&self.inverse_bind_scale)?;
        writer.write_quat::<LE>(&self.inverse_bind_rotation)?;

        let pos = writer.stream_position()?;
        writer.write_i32::<LE>((name_off as i64 - pos as i64) as i32)?;

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use approx::assert_ulps_eq;
    use glam::{vec3, Mat4, Quat};
    use io::SeekFrom;
    use std::io::Cursor;

    macro_rules! assert_joint_fuzzy_vec {
        ($a: ident, $b: ident, $field:tt) => {
            for (a, b) in $a.$field.to_array().iter().zip($b.$field.to_array().iter()) {
                assert_ulps_eq!(a, b);
            }
            $a.$field = $b.$field;
        };
    }
    macro_rules! assert_joint_fuzzy_mat {
        ($a: ident, $b: ident, $field:tt) => {
            for (a, b) in $a
                .$field
                .to_cols_array()
                .iter()
                .zip($b.$field.to_cols_array().iter())
            {
                assert_ulps_eq!(a, b);
            }
            $a.$field = $b.$field;
        };
    }

    #[test]
    fn roundtrip_write() {
        let mat = Mat4::from_scale_rotation_translation(
            vec3(0.5, 0.5, 0.5),
            Quat::from_array([1.0, 0.8, 0.1, 0.0]).normalize(),
            vec3(1.0, 50.0, 34.0),
        );

        let joint_name = "joint";
        let joint_size = 100; // size of joint in bytes
        let mut buf = Cursor::new(vec![0; joint_size + joint_name.len() + 1]); // +1 for nul terminator

        let mut a = Joint::new(
            joint_name.into(),
            1234,
            5432,
            9876,
            42.0,
            mat,
            mat.inverse(),
        );

        let name_off = joint_size as u64;

        a.to_writer(&mut buf, name_off).unwrap();

        buf.seek(SeekFrom::Start(name_off)).unwrap();
        buf.write_all(joint_name.as_bytes()).unwrap();
        buf.rewind().unwrap();

        println!("{buf:?}");

        let mut b = Joint::from_reader(&mut buf).unwrap();
        /*
         Because assert_eq isn't good with floats,
         we first check the float values with the 'approx' crate (see above macros),
         then *sets* the fields equal to each other,
         so that the assert_eq doesn't fail on those fields - but still checks everything else.
        */
        assert_joint_fuzzy_vec!(a, b, local_scale);
        assert_joint_fuzzy_vec!(a, b, local_rotation);
        assert_joint_fuzzy_vec!(a, b, local_translation);
        assert_joint_fuzzy_mat!(a, b, local_transform);

        assert_joint_fuzzy_vec!(a, b, inverse_bind_scale);
        assert_joint_fuzzy_vec!(a, b, inverse_bind_rotation);
        assert_joint_fuzzy_vec!(a, b, inverse_bind_translation);
        assert_joint_fuzzy_mat!(a, b, inverse_bind_transform);

        assert_eq!(a, b);
    }
}
