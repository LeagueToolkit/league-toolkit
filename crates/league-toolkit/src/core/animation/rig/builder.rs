use crate::core::animation::joint;

pub struct Builder {
    name: String,
    asset_name: String,
    root_joints: Vec<joint::Builder>,
}

impl Builder {
    pub fn new(name: impl Into<String>, asset_name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            asset_name: asset_name.into(),
            root_joints: vec![],
        }
    }
}