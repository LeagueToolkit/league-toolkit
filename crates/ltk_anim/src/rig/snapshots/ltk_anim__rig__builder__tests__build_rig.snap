---
source: crates/league-toolkit/src/core/animation/rig/builder.rs
expression: rig
---
RigResource {
    flags: 0,
    name: "my_rig",
    asset_name: "my_rig_asset",
    joints: [
        Joint {
            name: "root_1",
            flags: 10,
            id: 0,
            parent_id: -1,
            radius: 2.1,
            local_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            local_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            local_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            local_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
            inverse_bind_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            inverse_bind_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            inverse_bind_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            inverse_bind_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
        },
        Joint {
            name: "2a",
            flags: 11,
            id: 1,
            parent_id: 0,
            radius: 2.1,
            local_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            local_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            local_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            local_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
            inverse_bind_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            inverse_bind_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            inverse_bind_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            inverse_bind_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
        },
        Joint {
            name: "2b",
            flags: 12,
            id: 2,
            parent_id: 0,
            radius: 2.1,
            local_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            local_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            local_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            local_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
            inverse_bind_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            inverse_bind_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            inverse_bind_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            inverse_bind_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
        },
        Joint {
            name: "root_2",
            flags: 20,
            id: 3,
            parent_id: -1,
            radius: 2.1,
            local_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            local_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            local_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            local_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
            inverse_bind_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            inverse_bind_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            inverse_bind_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            inverse_bind_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
        },
        Joint {
            name: "2a",
            flags: 21,
            id: 4,
            parent_id: 3,
            radius: 2.1,
            local_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            local_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            local_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            local_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
            inverse_bind_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            inverse_bind_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            inverse_bind_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            inverse_bind_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
        },
        Joint {
            name: "2aa",
            flags: 211,
            id: 5,
            parent_id: 4,
            radius: 2.1,
            local_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            local_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            local_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            local_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
            inverse_bind_transform: Mat4 {
                x_axis: Vec4(
                    1.0,
                    0.0,
                    0.0,
                    0.0,
                ),
                y_axis: Vec4(
                    0.0,
                    1.0,
                    0.0,
                    0.0,
                ),
                z_axis: Vec4(
                    0.0,
                    0.0,
                    1.0,
                    0.0,
                ),
                w_axis: Vec4(
                    0.0,
                    0.0,
                    0.0,
                    1.0,
                ),
            },
            inverse_bind_translation: Vec3(
                0.0,
                0.0,
                0.0,
            ),
            inverse_bind_scale: Vec3(
                1.0,
                1.0,
                1.0,
            ),
            inverse_bind_rotation: Quat(
                0.0,
                0.0,
                0.0,
                1.0,
            ),
        },
    ],
    influences: [
        2,
        4,
    ],
}
